core/internal/runconsolelogs/linefile.go

```diff
@@ -0,0 +1,291 @@
+	cursor int64
+}
+
+// CreateLineFile creates an empty file or returns an error.
+//
+// The file must not already exist. Its parent directory must exist.
+func CreateLineFile(path string, perm fs.FileMode) (*lineFile, error) {
+	if err := os.WriteFile(string(path), []byte{}, perm); err != nil {
+		return nil, err
+	}
+
+	return &lineFile{path: path, nextLineNum: 0}, nil
+}
+
+// UpdateLines replaces lines in the file.
+//
+// This is written for performance when updating lines near the end of the file.
+func (f *lineFile) UpdateLines(lines sparselist.SparseList[string]) error {
+	if lines.Len() == 0 {
+		return nil
+	}
+
+	file, err := f.Open()
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	oldLinesReversed := []string{}
+	origNextLine := f.nextLineNum
+
+	// Pop off old lines until the line we're modifying.
+	for ; f.nextLineNum > lines.FirstIndex(); f.nextLineNum-- {
+		line, err := file.PopLine()
+		if err != nil {
+			return err
+		}
+
+		oldLinesReversed = append(oldLinesReversed, line)
+	}
+
+	// Write back lines, replacing old lines as necessary.
+loop:
+	for {
+		var lineToWrite string
+
+		newLine, hadNewLine := lines.Get(f.nextLineNum)
+		switch {
+		// Use the new line if there is one.
+		case hadNewLine:
+			lineToWrite = newLine
+
+		// Use the old line if there is one.
+		case f.nextLineNum < origNextLine:
+			idx := origNextLine - 1 - f.nextLineNum
+			lineToWrite = oldLinesReversed[idx]
+
+		// If we're adding lines past the original number of lines in the
+		// file, append blanks for any unspecified lines.
+		case f.nextLineNum < lines.LastIndex():
+			lineToWrite = ""
+
+		// No more lines to write, stop.
+		default:
+			break loop
+		}
+
+		f.nextLineNum++
+
+		if err := file.AppendLine(lineToWrite); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Open opens the file for modification.
+func (f *lineFile) Open() (*lineFileIO, error) {
+	file, err := os.OpenFile(f.path, os.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %v", err)
+	}
+
+	info, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to stat file: %v", err)
+	}
+
+	endOffset := info.Size()
+
+	return &lineFileIO{
+		f:           file,
+		bufStart:    endOffset,
+		buf:         make([]byte, 0),
+		nextLineNum: f.nextLineNum,
+		cursor:      endOffset,
+	}, nil
+}
+
+// Close flushes changes and closes the file handle.
+func (f *lineFileIO) Close() error {
+	if err := f.f.Truncate(f.bufStart + int64(len(f.buf))); err != nil {
+		return err
+	}
+
+	if err := f.f.Close(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// AppendLine writes a line to the file.
+//
+// The line must not contain '\n'.
+func (f *lineFileIO) AppendLine(line string) error {
+	lineStart := f.bufStart + int64(len(f.buf))
+
+	// Truncate the file to the current position first.
+	if len(f.buf) > 0 {
+		if err := f.f.Truncate(lineStart); err != nil {
+			return err
```

Comment:
when `f.buf` is truncated, it's being reset to an empty buffer, right? If `Truncate` or `WriteAt` fails, it might lead to inconsistencies or data loss...do we care?

Reply:
hmm.. good point. I'll move the state updates after the file modifications, so that we don't change any state on error.

Reply:
done!